demos/common/monitors/memory_monitor.cpp

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "memory_monitor.h"
+
+#ifdef WIN32
+#include <algorithm>
+
+#define PSAPI_VERSION 1 // for psapi
+#include <windows.h>
+#include <psapi.h>
+
+namespace {
+double getMemTotal() {
+    PERFORMANCE_INFORMATION performanceInformation;
+    if (!GetPerformanceInfo(&performanceInformation, sizeof(performanceInformation))) {
+        throw std::runtime_error("GetPerformanceInfo() failed");
+    }
+    return static_cast<double>(performanceInformation.PhysicalTotal * performanceInformation.PageSize)
+        / (1024 * 1024 * 1024);
+}
+double getSwapTotal() {
+    PERFORMANCE_INFORMATION performanceInformation;
+    if (!GetPerformanceInfo(&performanceInformation, sizeof(performanceInformation))) {
+        throw std::runtime_error("GetPerformanceInfo() failed");
+    }
+    return static_cast<double>(performanceInformation.CommitLimit * performanceInformation.PageSize)
```

Comment:
This is questionable. The commit limit is not the size of the swap file; it is the amount of memory that can be committed, which comprises both the size of the page file and the size of physical memory (minus the amount of memory that can't be paged out).

Reply:
Why minus the amount of memory that can't be paged out?
Should it be
```
usedSwap = performanceInformation.CommitTotal - (performanceInformation.PhysicalTotal - performanceInformation.PhysicalAvailable)
```
? And it looks I don't really need `swapTotal`

Reply:
> Why minus the amount of memory that can't be paged out?

Probably not the best wording on my part. What I mean is that some physical memory is reserved by the OS and isn't included in the commit limit (or the commit charge).

> usedSwap = performanceInformation.CommitTotal - (performanceInformation.PhysicalTotal - performanceInformation.PhysicalAvailable)

This seems close, but it's not quite correct. Imagine that a file is mapped into memory for writing and some changes have been made to it but not flushed to disk yet. The pages corresponding to this file will be counted in `PhysicalTotal`, but not in `PhysicalAvailable` or `CommitTotal`, which means they will affect `usedSwap`, even though they actually have no effect on swap. Same thing for the reserved pages I mentioned above.

Reply:
> The pages corresponding to this file will be counted in `PhysicalTotal`, but not in `PhysicalAvailable` or `CommitTotal`

https://answers.microsoft.com/en-us/windows/forum/windows_10-performance/physical-and-virtual-memory-in-windows-10/e36fb5bc-9ac8-49af-951c-e7d39b979938# claims:
> Committed (Virtual In Use) = Physical Memory in Use + Page file in Use

If the memory is counted in (PhysicalTotal - PhysicalAvailable) than it is counted in `CommitTotal`.

Reserved memory is also described there:
> Total Physical Memory = Installed - Hardware Reserved

So in memory_monitor's computations it is just not accounted anywhere.

Reply:
> If the memory is counted in (PhysicalTotal - PhysicalAvailable) than it is counted in CommitTotal.

Unfortunately, that is not true, because not all physical pages that are in use correspond to memory that is committed. I have actually written a program that creates the scenario I described above (with the unflushed changes to a mapped file) and by running that program, I got the memory stats to look like this:

```
CommitTotal: 6768248 K
CommitLimit: 10010272 K
PhysicalTotal: 8218664 K
PhysicalAvailable: 720016 K
```

… which, if you were to apply your formula, would yield negative swap usage. :-)

> Total Physical Memory = Installed - Hardware Reserved

"Hardware reserved" is not the same thing as "OS reserved". The former is regions of memory that are inaccessible because some device's registers are mapped there. The latter is regions of memory that contain critical code/data that must not ever be paged out. However, I added up the size of the page file and `PhysicalTotal` on my machine, and the result was exactly equal to `CommitLimit`, which suggests that OS-reserved RAM is included in the commit charge. So I was wrong on this point.

Reply:
BTW, if you want to repeat my experiments, here are the programs I used: https://gist.github.com/IRDonch/789ffef15945f08ec60696772f84d0cd

Reply:
Do you know a way to get size of used part of pageFile from C++? I can't find any.

How does it:
> I added up the size of the page file and PhysicalTotal on my machine, and the result was exactly equal to CommitLimit

proves that
>OS-reserved RAM is included in the commit charge

?

Reply:
> Do you know a way to get size of used part of pageFile from C++?

Not exactly, but in Performance Monitor there is a counter that shows the percentage of used page file space (I don't remember the exact name; I think it was under "Page file"). It's probably accessible through the PDH API.

> How does it: [...] proves that [...]?

The OS-reserved areas of RAM are included in `PhysicalTotal`, which is easy to check with [RAMMap](https://docs.microsoft.com/en-us/sysinternals/downloads/rammap) (for example, it shows that non-paged pool (which is one of these areas) is included in the total). Since the page file size plus `PhysicalTotal` exactly equals the commit limit, it means that the entirety of `PhysicalTotal` is included in the commit limit, including the OS-reserved areas. If the OS-reserved areas were not included in the commit charge, then the commit charge could never reach the commit limit, which would be silly.

Reply:
Thank you!
It looks that [PDH doesn't provide Physical Memory: Total](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/aa965225(v%3Dvs.85)#system-memory-performance-information), so in order to get used Paging File size I will need both PDH and `GetPerformanceInfo()` assuming that system state doesn't change significantly between collectiong info from these two sources.

Reply:
Updated